pdata/pprofile/merge.go

```diff
@@ -0,0 +1,497 @@
+
+	for i := 0; i < sourceLinks.Len(); i++ {
+		sourceLink := sourceLinks.At(i)
+
+		sample := NewSample()
+		if err := SetLink(destLinks, sample, sourceLink); err != nil {
+			return fmt.Errorf("failed to merge link table entry at index %d: %w", i, err)
+		}
+
+		idx, err := safeInt32(i)
+		if err != nil {
+			return fmt.Errorf("link index: %w", err)
+		}
+		linkMapping[idx] = sample.LinkIndex()
+	}
+	return nil
+}
+
+func (ms Profiles) mergeAttributeTable(destDict, sourceDict ProfilesDictionary, attributeMapping, stringMapping map[int32]int32) error {
+	destAttributes := destDict.AttributeTable()
+	sourceAttributes := sourceDict.AttributeTable()
+
+	for i := 0; i < sourceAttributes.Len(); i++ {
+		sourceAttr := sourceAttributes.At(i)
+
+		keyIdx := sourceAttr.KeyStrindex()
+		if keyIdx < 0 || int(keyIdx) >= sourceDict.StringTable().Len() {
+			return fmt.Errorf("attribute key index %d out of range for entry %d", keyIdx, i)
+		}
+		key := sourceDict.StringTable().At(int(keyIdx))
+		profile := NewProfile()
+		if err := PutAttribute(destAttributes, profile, destDict, key, sourceAttr.Value()); err != nil {
+			return fmt.Errorf("failed to merge attribute table entry at index %d: %w", i, err)
+		}
+
+		if profile.AttributeIndices().Len() == 0 {
+			return fmt.Errorf("attribute indices not updated for entry %d", i)
+		}
+
+		mappedIdx := profile.AttributeIndices().At(profile.AttributeIndices().Len() - 1)
+
+		if unitIdx := sourceAttr.UnitStrindex(); unitIdx >= 0 {
+			if newUnitIdx, exists := stringMapping[unitIdx]; exists {
+				destAttributes.At(int(mappedIdx)).SetUnitStrindex(newUnitIdx)
+			}
+		}
+
+		idx, err := safeInt32(i)
+		if err != nil {
+			return fmt.Errorf("attribute index: %w", err)
+		}
+		attributeMapping[idx] = mappedIdx
+	}
+	return nil
+}
+
+func (ms Profiles) mergeResourceProfiles(sourceResourceProfiles ResourceProfilesSlice, mappings *indexMappings) error {
+	destResourceProfiles := ms.ResourceProfiles()
+
+	for i := 0; i < sourceResourceProfiles.Len(); i++ {
+		sourceRP := sourceResourceProfiles.At(i)
+		destRP := destResourceProfiles.AppendEmpty()
+
+		sourceRP.Resource().CopyTo(destRP.Resource())
+		destRP.SetSchemaUrl(sourceRP.SchemaUrl())
+
+		ms.mergeScopeProfiles(sourceRP.ScopeProfiles(), destRP.ScopeProfiles(), mappings)
+	}
+
+	return nil
+}
+
+func (ms Profiles) mergeScopeProfiles(sourceScopeProfiles, destScopeProfiles ScopeProfilesSlice, mappings *indexMappings) {
+	for i := 0; i < sourceScopeProfiles.Len(); i++ {
+		sourceSP := sourceScopeProfiles.At(i)
+		destSP := destScopeProfiles.AppendEmpty()
+
+		sourceSP.Scope().CopyTo(destSP.Scope())
+		destSP.SetSchemaUrl(sourceSP.SchemaUrl())
+
+		ms.mergeProfiles(sourceSP.Profiles(), destSP.Profiles(), mappings)
+	}
+}
+
+func (ms Profiles) mergeProfiles(sourceProfiles, destProfiles ProfilesSlice, mappings *indexMappings) {
+	for i := 0; i < sourceProfiles.Len(); i++ {
+		sourceProfile := sourceProfiles.At(i)
+		destProfile := destProfiles.AppendEmpty()
+
```

Comment:
Profiles should not get merged without further checks.
E.g. merging different sample types will invalidate all data. Also, for investigations, the timestamps are crucial. Here the destination timestamp is overwritten with the source timestamps. And similar things happen to other fields of the message.